external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java

```diff
@@ -280,61 +280,62 @@ public Builder(String bootstrapServers, Subscription subscription) {
+                    (RecordTranslator<K, V>) builder.translator, new HashMap<String, Object>());
+
+            this.kafkaProps.putAll(builder.kafkaProps);
             this.pollTimeoutMs = builder.pollTimeoutMs;
             this.offsetCommitPeriodMs = builder.offsetCommitPeriodMs;
             this.firstPollOffsetStrategy = builder.firstPollOffsetStrategy;
             this.maxUncommittedOffsets = builder.maxUncommittedOffsets;
-            //this could result in a lot of class case exceptions at runtime,
-            // but because some translators will work no matter what the generics
-            // are I thought it best not to force someone to reset the translator
-            // when they change the key/value types.
-            this.translator = (RecordTranslator<K, V>) builder.translator;
             this.retryService = builder.retryService;
 
-            if (keyDes != null) {
-                this.kafkaProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, keyDes.getClass());
-            } else if (keyDesClazz != null) {
-                this.kafkaProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, keyDesClazz);
-            } else if (valueDes != null) {
-                this.kafkaProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, valueDes.getClass());
-            } else if (valueDesClazz != null) {
-                this.kafkaProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, valueDesClazz);
-            }
+            setSerDeKafkaProps(keyDes, keyDesClazz, valueDes, valueDesClazz);
+        }
 
+        private Builder(SerializableDeserializer<K> keyDes, Class<? extends Deserializer<K>> keyDesClazz,
+               SerializableDeserializer<V> valueDes, Class<? extends Deserializer<V>> valueDesClazz,
+                       Subscription subscription, RecordTranslator<K, V> translator, Map<String, Object> kafkaProps) {
             this.keyDes = keyDes;
             this.keyDesClazz = keyDesClazz;
             this.valueDes = valueDes;
             this.valueDesClazz = valueDesClazz;
+            this.subscription = subscription;
+            this.translator = translator;
+            this.kafkaProps = kafkaProps;
+        }
+
+        private void setSerDeKafkaProps(SerializableDeserializer<K> keyDes, Class<? extends Deserializer<K>> keyDesClazz, SerializableDeserializer<V> valueDes, Class<? extends Deserializer<V>> valueDesClazz) {
+            setNonNullValProp(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, keyDesClazz, false);
+            setNonNullValProp(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, keyDes, true);
+            setNonNullValProp(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, valueDesClazz, false);
+            setNonNullValProp(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, valueDes, true);
+        }
+
+        private void setNonNullValProp(String key, Object val, boolean isClass) {
+            if (val != null) {
+                if (isClass) {
```

Comment:
I think you should rename the parameter. It's true if the val is not a class.

Reply:
I removed this method